File: scripts/recover-login.js
/**
 * recover-login.js
 * Lists all users in the local Electron DB and resets the password for
 * the first admin account (or a specific email you set below).
 *
 * Usage:
 *   node scripts/recover-login.js
 *   node scripts/recover-login.js <email> <newpassword>
 *
 * It automatically finds the Electron userData DB path.
 */
'use strict';

const path  = require('path');
const fs    = require('fs');
const bcrypt = require('bcryptjs');

// ─── Resolve DB path (same logic as server/db.js) ───────────────────────────
const ELECTRON_USER_DATA = path.join(
  process.env.APPDATA || path.join(require('os').homedir(), 'AppData', 'Roaming'),
  'ai-content-studio'
);
const DB_PATH = path.join(ELECTRON_USER_DATA, 'data', 'saas.db');

if (!fs.existsSync(DB_PATH)) {
  console.error('❌ DB not found at:', DB_PATH);
  process.exit(1);
}

const Database = require('better-sqlite3');
const db = new Database(DB_PATH);
db.pragma('journal_mode = WAL');

// ─── Show all users ───────────────────────────────────────────────────────────
const users = db.prepare('SELECT id, email, name, is_admin, is_owner, verified FROM users').all();
console.log('\n📋 Users in local DB:');
console.log('─'.repeat(70));
users.forEach((u, i) => {
  const role = u.is_owner ? '👑 OWNER' : u.is_admin ? '🔑 ADMIN' : '👤 user';
  const ver  = u.verified  ? '✅' : '❌ unverified';
  console.log(`${i + 1}. ${u.email} (${u.name}) ${role} ${ver}`);
});
console.log('─'.repeat(70));

if (users.length === 0) {
  console.log('\n⚠️  No users found. The DB might be empty or on wrong path.');
  db.close();
  process.exit(0);
}

// ─── Password reset ────────────────────────────────────────────────────────
const [,, targetEmail, newPassword] = process.argv;

async function resetPassword(email, newPwd) {
  const user = db.prepare('SELECT id FROM users WHERE email = ?').get(email.toLowerCase());
  if (!user) {
    console.error(`\n❌ No user found with email: ${email}`);
    db.close();
    process.exit(1);
  }
  const hash = await bcrypt.hash(newPwd, 12);
  db.prepare('UPDATE users SET password_hash = ?, verified = 1, reset_token = NULL WHERE id = ?').run(hash, user.id);
  console.log(`\n✅ Password reset for: ${email}`);
  console.log(`   New password: ${newPwd}`);
  console.log('\n   You can now log in to the app with these credentials.\n');
  db.close();
}

if (targetEmail && newPassword) {
  resetPassword(targetEmail, newPassword);
} else if (users.length === 1) {
  // Auto-target single account
  const autoEmail = users[0].email;
  const autoPwd   = 'Kyros2026!';
  console.log(`\n🔧 Single user found. Auto-resetting password for: ${autoEmail}`);
  console.log(`   New temporary password: ${autoPwd}`);
  resetPassword(autoEmail, autoPwd);
} else {
  console.log('\nTo reset a password, run:');
  console.log('  node scripts/recover-login.js <email> <newpassword>');
  console.log('\nExample:');
  console.log(`  node scripts/recover-login.js ${users[0].email} MyNewPassword123`);
  db.close();
}

File: scripts/recover_login.py
"""
recover_login.py -- Lists users and resets password in Kyros Studio local DB.
No emojis used -- safe on Windows cp1252 terminals.

Usage:
    python scripts/recover_login.py
    python scripts/recover_login.py <email> <new_password>
"""
import sys
import os
import sqlite3
import subprocess

# Force UTF-8 output on Windows
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

DB_PATH = os.path.join(
    os.environ.get('APPDATA', os.path.join(os.path.expanduser('~'), 'AppData', 'Roaming')),
    'ai-content-studio', 'data', 'saas.db'
)

if not os.path.exists(DB_PATH):
    print("ERROR: DB not found at:", DB_PATH)
    sys.exit(1)

print("OK - Found DB at:", DB_PATH)
print("   Size:", f"{os.path.getsize(DB_PATH):,}", "bytes")

conn = sqlite3.connect(DB_PATH)
conn.row_factory = sqlite3.Row

rows = conn.execute(
    'SELECT id, email, name, is_admin, is_owner, verified FROM users ORDER BY created_at'
).fetchall()

print("\n=== Users in local DB ===")
print("-" * 70)
for i, u in enumerate(rows, 1):
    role = "OWNER" if u['is_owner'] else ("ADMIN" if u['is_admin'] else "user")
    ver  = "verified" if u['verified'] else "UNVERIFIED"
    print(f"  {i}. {u['email']}  ({u['name']})  [{role}]  [{ver}]")
print("-" * 70)

if not rows:
    print("\nWARNING: No users found.")
    conn.close()
    sys.exit(0)


def make_bcrypt_hash(password, rounds=12):
    """Install bcrypt if needed, then hash the password."""
    try:
        import bcrypt as _bcrypt
    except ImportError:
        print("Installing bcrypt package...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", "bcrypt", "-q"])
        import bcrypt as _bcrypt
    hashed = _bcrypt.hashpw(password.encode(), _bcrypt.gensalt(rounds))
    return hashed.decode()


def reset_password(email, new_password):
    user = conn.execute(
        'SELECT id FROM users WHERE LOWER(email) = LOWER(?)', (email,)
    ).fetchone()
    if not user:
        print("\nERROR: No user found with email:", email)
        conn.close()
        sys.exit(1)

    print("\nHashing new password (this takes a few seconds)...")
    pw_hash = make_bcrypt_hash(new_password)
    conn.execute(
        'UPDATE users SET password_hash = ?, verified = 1, reset_token = NULL, reset_token_expiry = NULL WHERE id = ?',
        (pw_hash, user['id'])
    )
    conn.commit()
    print("\n>>> Password reset OK <<<")
    print("    Email:        ", email)
    print("    New password: ", new_password)
    print("\n    Open Kyros Studio and log in with these credentials.\n")


args = sys.argv[1:]
if len(args) == 2:
    reset_password(args[0], args[1])
elif len(rows) == 1:
    auto_email = rows[0]['email']
    auto_pwd   = "Kyros2026!"
    print("\nSingle user found -- auto-resetting:", auto_email)
    reset_password(auto_email, auto_pwd)
else:
    print("\nTo reset a password run:")
    print("  python scripts/recover_login.py <email> <newpassword>")
    print("\nExample:")
    print(f"  python scripts/recover_login.py {rows[0]['email']} Kyros2026!")
    conn.close()
